src/ssl.c

```diff
@@ -48370,6 +48370,357 @@ int wolfSSL_PEM_write_bio_PKCS7(WOLFSSL_BIO* bio, PKCS7* p7)
+    if (!curHeader) WOLFSSL_MSG("Content-Transfer-Encoding header not found, assuming base64 encoding.");
+    else if (XSTRNCMP(curHeader->body, "base64", XSTRLEN("base64"))) {
+        WOLFSSL_MSG("S/MIME encodings other than base64 not currently supported.\n");
+        return NULL;
+    }
+
+    if (!section || sectionLen <= 0) return NULL;
+    /* Strip trailing newlines. */
+    while ((section[sectionLen-1] == '\r' || section[sectionLen-1] == '\n') && sectionLen > 0) sectionLen--;
+    section[sectionLen] = '\0';
+
+    word32 outLen = ((sectionLen*3+3)/4)+1;
+    byte* out = XMALLOC(outLen*sizeof(byte), NULL, DYNAMIC_TYPE_PKCS7);
+    if (!out) return NULL;
+    byte* outHead = out;
+    ret = Base64_Decode((const byte*)section, sectionLen, out, &outLen);
+    if (ret < 0) {
+        WOLFSSL_MSG("Error base64 decoding S/MIME message.\n");
+        return NULL;
+    }
+    pkcs7 = wolfSSL_d2i_PKCS7(NULL, (const unsigned char**)&out, outLen);
+
+
+    MIME_free_headers(allHeaders);
+    if (outHead) XFREE(outHead, NULL, DYNAMIC_TYPE_PKCS7);
+    if (section) XFREE(section, NULL, DYNAMIC_TYPE_PKCS7);
+
+
+    return pkcs7;
+}
+
+int MIME_parse_headers(WOLFSSL_BIO* in, int inLen, MimeHeader** headers)
+{
+    MimeHeader *nextHeader = XMALLOC(sizeof(MimeHeader), NULL, DYNAMIC_TYPE_PKCS7);
+    MimeHeader *curHeader = NULL;
+    MimeParameter *nextParam = XMALLOC(sizeof(MimeParameter), NULL, DYNAMIC_TYPE_PKCS7);
+    word32 start = 0;
+    word32 end = 0;
+    char *nameAttr = NULL;
+    char *bodyVal = NULL;
+    MimeTypes mimeType = MIME_HEADER;
+    MimeStatus mimeStatus = MIME_NAMEATTRIB;
+    int ret = 0;
+    int pos = 0;
+    int lineLen = 0;
+    byte isEnd = 0;
+    char* lines = XMALLOC((inLen+1)*sizeof(char), NULL, DYNAMIC_TYPE_PKCS7);
+    char* curLine = lines;
+
+    if (!nextHeader || !nextParam || !lines) return MEMORY_E;
+    if (inLen < 0) return ASN_PARSE_E;
+    XMEMSET(nextHeader, 0, sizeof(MimeHeader));
+    XMEMSET(nextParam, 0, sizeof(MimeParameter));
+
+    lineLen = wolfSSL_BIO_gets(in, curLine, inLen);
+    if (lineLen<=0) return ASN_PARSE_E;
+    inLen -= lineLen;
+    if ((lineLen==2 && !XSTRNCMP(curLine, "\r\n", 2)) ||
+        (lineLen==1 && (curLine[0] == '\r' || curLine[0] == '\n')))
+        isEnd = 1;
+
+    while (!isEnd) {
+        // Leftover from previous line, add parameters to previous header.
+        if (curLine[0] == ' ' && curHeader) mimeType = MIME_PARAMETER;
+        else mimeType = MIME_HEADER;
+        start = end = 0;
+
+        for (pos = 0; pos < lineLen; pos++) {
+            char cur = curLine[pos];
+
+            if (mimeStatus == MIME_NAMEATTRIB &&
+                ((cur == ':' && mimeType == MIME_HEADER) || (cur == '=' && mimeType == MIME_PARAMETER))) {
+                mimeStatus = MIME_BODYVALUE;
+                end = pos-1;
+                ret = MIME_header_strip(curLine, &nameAttr, start, end);
+                if (ret) return ret;
```

Comment:
Please double check we are freeing all local memory when returning from function.